mlflow/utils/autologging/logging_and_warnings.py

```diff
@@ -0,0 +1,265 @@
+    MLflow's autologging implementation. MLflow warnings are also redirected from `sys.stderr`
+    to an MLflow logger with level WARNING. Warnings issued by code outside of MLflow are
+    not modified. When the context manager exits, the original output behavior for MLflow warnings
+    is restored.
+
+    Note that the implementation of `augment_mlflow_warnings` is *not* threadsafe.
+    """
+    try:
+        _WARNINGS_CONTROLLER.set_global_mlflow_warnings_rerouting_state(rerouted=True)
+        yield
+    finally:
+        _WARNINGS_CONTROLLER.set_global_mlflow_warnings_rerouting_state(rerouted=False)
+
+
+@contextmanager
+def silence_warnings_and_mlflow_event_logs_if_necessary(autologging_integration):
+    """
+    Context manager that silences all warnings (MLflow warnings and others) and all MLflow event
+    logging statements upon entry if silent mode is enabled for the specified
+    `autologging_integration` (i.e. if the `silent` attribute of the integration's configuration is
+    `True`). If silent mode is *not* enabled for the specified integration, this context manager is
+    a no-op. Upon exit, this context manager re-enables warnings and MLflow event logging
+    statements, making a best-effort attempt to provide thread safety (details below).
+
+    When the context manager is active, MLflow event logging statements are silenced *globally
+    across all threads*, while warnings (MLflow warnings and others) are silenced for the
+    *current thread*. The reasons for this behavior are as follows:
+
+    - During autologging sessions with `silent=True`, all MLflow warnings and MLflow logging
+      statements should be silenced in a threadsafe fashion. We should not leak MLflow logging
+      statements in multithreaded contexts.
+
+    - During autologging sessions with `silent=True`, non-MLflow warnings raised by autologging
+      implementation code in MLflow should be suppressed.
+
+    - During autologging sessions with `silent=True`, non-MLflow warnings and event logging
+      statements emitted from original / underlying ML routines should never be suppressed.
+      We should not omit any of these warnings and event logging statements in multithreaded
+      contexts.
+
+    This context manager attempts to be threadsafe across autologging sessions, with the following
+    caveats:
+
+    - If a silent autologging session is run in parallel with a non-silent autologging session,
+      MLflow event logging statements and warnings from the non-silent session may be suppressed.
+
+    - If `warnings.showwarning` is unexpectedly re-assigned during the execution of a user's
+      original / underlying ML code within an autologging session, warnings may not be
+      properly suppressed.
+
+    - If MLflow's autologging implementation code makes calls to multithreaded APIs that raise warnings
+      from a different thread, these warnings may not be properly suppressed.
```

Comment:
I'd recommend starting with this comment for context about threading behavior. Please let me know if you have any questions!